File: drills/python/solutions/nc125_reverse_linked_list__by_recursion.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reverseList(head):
    # Reverse the rest, then hook the current node onto its end. The new head
    # comes back unchanged through every frame, which is why it is returned
    # rather than tracked -- and head.next.next = head is the whole rewiring,
    # done on the way back out. O(n) stack, against the loop's O(1).
    if head is None or head.next is None:
        return head
    reversed_rest = reverseList(head.next)
    head.next.next = head
    head.next = None
    return reversed_rest

File: drills/ts/solutions/nc125_reverse_linked_list__by_recursion.ts
export class ListNode {
  val: number;
  next: ListNode | null;

  constructor(val = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

export function reverseList(head: ListNode | null): ListNode | null {
  // Reverse the rest, then hook the current node onto its end. The new head
  // comes back unchanged through every frame, which is why it is returned
  // rather than tracked -- and head.next.next = head is the whole rewiring,
  // done on the way back out. O(n) stack, against the loop's O(1).
  if (head === null || head.next === null) return head;
  const reversedRest = reverseList(head.next);
  head.next.next = head;
  head.next = null;
  return reversedRest;
}
